internal/expr/prepare.go

```diff
@@ -145,54 +123,127 @@ func prepareOutput(ti typeNameToInfo, p *outputPart) (outCols []fullName, typeMe
 					return nil, nil, err
 				}
 				typeMembers = append(typeMembers, tm)
-				outCols = append(outCols, fullName{pref, t.name})
+				genCols = append(genCols, fullName{pref, t.name})
 			}
 		}
-		return outCols, typeMembers, nil
+		return genCols, typeMembers, nil
 	} else if numColumns > 1 && starColumns > 0 {
 		return nil, nil, fmt.Errorf("invalid asterisk in columns")
 	}
 
-	// Case 2: Explicit columns, single asterisk type e.g. "(col1, t.col2) AS &P.*".
+	// Case 2: Explicit columns, single asterisk type.
+	// For example the expressions:
+	//  "(col1, col2) VALUES ($P.*)"
+	//  "(col1, t.col2) AS (&P.*)"
 	if starTypes == 1 && numTypes == 1 {
-		if info, err = fetchInfo(p.targetTypes[0].prefix); err != nil {
+		info, err := ti.lookupInfo(types[0].prefix)
+		if err != nil {
 			return nil, nil, err
 		}
-		for _, c := range p.sourceColumns {
+		for _, c := range columns {
 			tm, err := info.typeMember(c.name)
 			if err != nil {
 				return nil, nil, err
 			}
 			typeMembers = append(typeMembers, tm)
-			outCols = append(outCols, c)
+			genCols = append(genCols, c)
 		}
-		return outCols, typeMembers, nil
+		return genCols, typeMembers, nil
 	} else if starTypes > 0 && numTypes > 1 {
 		return nil, nil, fmt.Errorf("invalid asterisk in types")
 	}
 
-	// Case 3: Explicit columns and types e.g. "(col1, col2) AS (&P.name, &P.id)".
+	// Case 3: Explicit columns and types.
+	// For example the expressions:
+	//  "(col1, col2) VALUES ($P.name, $A.id)"
+	//  "(col1, col2) AS (&P.name, &P.id)"
 	if numColumns == numTypes {
-		for i, c := range p.sourceColumns {
-			t := p.targetTypes[i]
-			if info, err = fetchInfo(t.prefix); err != nil {
+		for i, c := range columns {
+			t := types[i]
```

Comment:
[Nit] Can we rename this to `typeNames`, I was not expecting `types[i]` to resolve to a `fullName`.

Reply:
Yea, will change that.